### PR TITLE
fix: improve the processing logic of generic functions

### PR DESCRIPTION
**Pull request by Shanks0224**

### CODE REVIEW

File: doc/developer-guide/generic_type.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    constructor(x: T) {
+        this._x = x;
+    }
+    bar<T>(a: T) {
+        console.log('bar: ' + a);
+    }
+}
+```
+
+#### Creation of class type
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:19]:
```suggestion
#### Creation of generic class
```

File: doc/developer-guide/generic_type.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+![](./img/generic_class_type_create.jpg)
+
+The steps to create a generic class type with type arguments are as follows:
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:23]:
```suggestion
The steps to create a generic class with generic type are as follows:
```

File: doc/developer-guide/generic_type.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+- create class scope
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:24]: I think that the pipeline of creating generic class is same as creating normal class, what needed to be recorded is what's the difference of `TypeParameter` and other normal type.

Eg, how to store the `T`, and how to ensure that all `T` in the generic class is the same one.

File: doc/developer-guide/generic_type.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+- create class scope  
+  &ensp;The compiler will create a `ClassScope` for each generic class and a `FunctionScope` for each method.
+- create class type  
+  &ensp;Create a generic class type and method types based on `classDeclarationNode`, and associate them with the created `ClassScope` and `FunctionScope`s.
+- create parameters and local variables of method  
+  &ensp;According to `classDeclarationNode`, create parameters and local variables for each method, and save them in the corresponding `FunctionScope`.
+- create the function body of method  
+  &ensp;According to `classDeclarationNode`, create a function body for each method and save it in the corresponding `FunctionScope`.
+
+#### Specialization of class type
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:33]:
```suggestion
#### Creation of specialized class
```

File: doc/developer-guide/generic_type.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+The creation of the specialized class type needs to be based on the `classDeclarationNode` information of the generic class type.
+
+![](./img/generic_class_type_specialize.jpg)
+
+Use `new` to create a new specialized class type:
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:38]: Had better record when we will create the specialized class. 

When we meet `NewExpression`, we will know that a new specialized class should be created, which includes new specialized class scope information. 
However, I remind that the new specialized class scope will be created in `semantic tree`? not in `ExpressionProcessor`?

Why put this step into semantic tree? If there any chance that we can done the specialization in frontend?

File: doc/developer-guide/generic_type.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    _x: number;
+
+    constructor(x: number) {
+        this._x = x;
+    }
+    bar<T>(a: T) {
+        console.log('bar: ' + a);
+    }
+}
+```
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:60]: In this situation, how many class scopes should we create in total?

File: doc/developer-guide/generic_type.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  &ensp;According to `classDeclarationNode`, create parameters and local variables for each method, and save them in the corresponding `FunctionScope`.
+- create the function body of method  
+  &ensp;According to `classDeclarationNode`, create a function body for each method and save it in the corresponding `FunctionScope`.
+
+#### Specialization of method type
+
+![](./img/generic_class_without_typeArgs_type_specialize.jpg)
+
+When a generic method is called, a specialization of the function is performed:
+- create new method scope
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:81]: So in this situation, we only need to create a new specialized function scope, no need to create a new specialized class scope, right?

File: doc/developer-guide/generic_type.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+When a generic method is called, a specialization of the function is performed:
+- create new method scope  
+ Copy the `FunctionScope` of the generic method and specialize it.
+- create new method type  
+ Copy the type of the generic method and specialize it.
+- specialize parameters and local variables  
+ Specialize the parameters and local variables in the new `FunctionScope`.
+- specialize the method body  
+ Specialize method body in the new `FunctionScope`.
+
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:89]: I suggest that put a summarized picture after each case, which will introduce the whole scope relationship, like:

                             GlobalScope
                              /              \
     ClassScope_Generic    ClassScope_Number
             /    \                                 /    \
           ...       ...                           ...    ...

File: doc/developer-guide/generic_type.md (hunk 7 above)
Comment (yviansu) [inline doc/developer-guide/generic_type.md:81]: Still, in the `CallExpression`: `a.bar(1)`, we can know that a new specialized function scope should be created, is this new scope creation pipeline is invoked in `ExpressionProcessor`? If not, why not? And where we create the new specialized scope?

File: doc/developer-guide/generic_type.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+``` TypeScript
+function foo<T>(a: T) {
+    console.log(a);
+}
+
+foo(1);
+foo('hello');
+```
+
+### Creation of function type
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:101]:
```suggestion
### Creation of generic function
```

File: doc/developer-guide/generic_type.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+- create function scope  
+  &ensp;The compiler will create a `FunctionScope` for each `FunctionDeclarationNode`.
+- create function type  
+  &ensp;According to the `FunctionDeclarationNode`, create a new function type and associate it with the created `FunctionScope`.
+- create parameters and local variables of the function  
+  &ensp;According to `FunctionDeclarationNode`, create parameters and local variables, and save them in the corresponding `FunctionScope`.
+- create the function body  
+  &ensp;According to `FunctionDeclarationNode`, create the function body and save it in the corresponding `FunctionScope`.
+
+### Specialization of function type
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:115]:
```suggestion
### Creation of specialized function
```

File: doc/developer-guide/generic_type.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    constructor(x: number) {
+        this._x = x;
+    }
+    bar<T>(a: T) {
+        console.log('bar: ' + a);
+    }
+}
+```
+
+#### Creation of class type
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:62]: same as above

File: doc/developer-guide/generic_type.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+- create class scope  
+  &ensp;The compiler will create a `ClassScope` for each generic class and a `FunctionScope` for each method.
+- create class type  
+  &ensp;Create a generic class type and method types based on `classDeclarationNode`, and associate them with the created `ClassScope` and `FunctionScope`s.
+- create parameters and local variables of method  
+  &ensp;According to `classDeclarationNode`, create parameters and local variables for each method, and save them in the corresponding `FunctionScope`.
+- create the function body of method  
+  &ensp;According to `classDeclarationNode`, create a function body for each method and save it in the corresponding `FunctionScope`.
+
+#### Specialization of method type
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:76]:
```suggestion
#### Creation of specialized method
```

File: doc/developer-guide/generic_type.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+  &ensp;According to `FunctionDeclarationNode`, create parameters and local variables, and save them in the corresponding `FunctionScope`.
+- create the function body  
+  &ensp;According to `FunctionDeclarationNode`, create the function body and save it in the corresponding `FunctionScope`.
+
+### Specialization of function type
+
+![](./img/generic_function_type_specialize.jpg)
+
+When a generic function is called, a specialization of the function will be performed:
+- copy function scope
```
Comment (yviansu) [inline doc/developer-guide/generic_type.md:120]: When will perform this operation?

File: doc/developer-guide/img/generic_class_type_specialize.jpg
Comment (yviansu) [inline doc/developer-guide/img/generic_class_type_specialize.jpg:1]: I suggest that the pipeline should also combined with our compilation pipeline.

For example, we can use horizontal lines to represent the steps of the compilation:
->
--------------------Frontend--------------------SemanticTree--------------------Backend--------------------
--scopeScanner--typeResolver--....

And we represent the current step of specialization vertically:
|         create scope    
|                   type specilaztion
|              ...
|            ...

File: doc/developer-guide/generic_type.md (hunk 1 above)
Comment (Shanks0224) [reply doc/developer-guide/generic_type.md]: I will delete the document first. After the document modification is completed, I will create a new PR and submit it to the internal Repo.

**Comment by xujuntwt95329** [comment]
@Shanks0224  Thanks for the great work!

**Review by xujuntwt95329 — APPROVED** [review]
LGTM

**Pull request merged**